Framework/DataHandling/src/H5Util.cpp

```diff
@@ -0,0 +1,306 @@
+    return readString(data);
+  } catch (H5::FileIException &e) {
+    UNUSED_ARG(e);
+    return "";
+  } catch (H5::GroupIException &e) {
+    UNUSED_ARG(e);
+    return "";
+  }
+}
+
+std::string readString(H5::Group &group, const std::string &name) {
+  try {
+    auto data = group.openDataSet(name);
+    return readString(data);
+  } catch (H5::GroupIException &e) {
+    UNUSED_ARG(e);
+    return "";
+  }
+}
+
+std::string readString(H5::DataSet &dataset) {
+  std::string value;
+  dataset.read(value, dataset.getDataType(), dataset.getSpace());
+  return value;
+}
+
+template <typename NumT>
+std::vector<NumT> readArray1DCoerce(H5::Group &group, const std::string &name) {
+  std::vector<NumT> result;
+
+  try {
+    DataSet dataset = group.openDataSet(name);
+    result = readArray1DCoerce<NumT>(dataset);
+  } catch (H5::GroupIException &e) {
+    UNUSED_ARG(e);
+    g_log.information() << "Failed to open dataset \"" << name << "\"\n";
+  } catch (H5::DataTypeIException &e) {
+    UNUSED_ARG(e);
+    g_log.information() << "DataSet \"" << name << "\" should be double"
+                        << "\n";
+  }
+
+  return result;
+}
+
+namespace {
+// correct for widening - narrowing should use something else
+template <typename InputNumT, typename OutputNumT>
+std::vector<OutputNumT> wideningRead(DataSet &dataset,
+                                     const DataType &dataType) {
+  DataSpace dataSpace = dataset.getSpace();
+
+  std::vector<InputNumT> temp(dataSpace.getSelectNpoints());
+  dataset.read(&temp[0], dataType, dataSpace);
+
+  std::vector<OutputNumT> result;
+  result.resize(temp.size());
+
+  std::transform(temp.begin(), temp.end(), result.begin(),
+                 [](const InputNumT a) { // lambda
+                   return static_cast<OutputNumT>(a);
+                 });
+
+  return result;
+}
+
+// should be correct for widening too
+template <typename InputNumT, typename OutputNumT>
+std::vector<OutputNumT> narrowingRead(DataSet &dataset,
+                                      const DataType &dataType) {
+  DataSpace dataSpace = dataset.getSpace();
+
+  std::vector<InputNumT> temp(dataSpace.getSelectNpoints());
+  dataset.read(&temp[0], dataType, dataSpace);
+
+  std::vector<OutputNumT> result;
+  result.resize(temp.size());
+
+  std::transform(temp.begin(), temp.end(), result.begin(),
+                 [](const InputNumT a) { // lambda
+                   return boost::numeric_cast<OutputNumT>(a);
+                 });
+
+  return result;
+}
+} // anonymous
+
+template <typename NumT> std::vector<NumT> readArray1DCoerce(DataSet &dataset) {
+  DataType dataType = dataset.getDataType();
+
+  if (getType<NumT>() == dataType) { // no conversion necessary
+    std::vector<NumT> result;
+    DataSpace dataSpace = dataset.getSpace();
+    result.resize(dataSpace.getSelectNpoints());
+    dataset.read(&result[0], dataType, dataSpace);
+    return result;
+  }
+
+  if (PredType::NATIVE_INT32 == dataType) {
```

Comment:
Can/should we use something like [narrow](https://github.com/Microsoft/GSL/blob/a9f865900d28b854de5ead971aadb82e5ef9ed40/include/gsl_util.h#L104) (also see [this](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#es46-avoid-lossy-narrowing-truncating-arithmetic-conversions)) in the GSL instead of setting narrowing or widening "by hand"?